src/db.c

```diff
@@ -792,16 +792,32 @@ void keysCommand(client *c) {
     setDeferredArrayLen(c,replylen,numkeys);
 }
 
+/* Data used by the dict scan callback. */
+typedef struct {
+    list *keys;   /* elements that collect from dict*/
+    robj *o;      /* o must be a hash/set/zset object, NULL means current db*/
+    sds typename; /* the particular type when scan the db */
+} scanData;
+
 /* This callback is used by scanGenericCommand in order to collect elements
  * returned by the dictionary iterator into a list. */
 void scanCallback(void *privdata, const dictEntry *de) {
-    void **pd = (void**) privdata;
-    list *keys = pd[0];
-    robj *o = pd[1];
+    scanData *data = (scanData *)privdata;
+    list *keys = data->keys;
+    robj *o = data->o;
     robj *key, *val = NULL;
 
+    /* typename is only meaningful when o is NULL*/
+    serverAssert(!data->typename || !o);
+
     if (o == NULL) {
         sds sdskey = dictGetKey(de);
+        robj *val = dictGetVal(de);
+        /* Filter an element if it isn't the type we want. */
+        if (data->typename) {
+            char *type = getObjectTypeName(val);
+            if (strcasecmp((char *)data->typename, type)) return;
+        }
```

Comment:
there's a reason all the filtering (pattern matching, expiry, etc), is done after fetching all the keys.
with this change, the loop that's suppose to stop at `count` can take very long time if it doesn't find anything.
it'll possibly return better results to the caller, but at a latency cost.
if we wanna move the filtering to the scan callback, we must also increment some counter so that we can stop the loop after the right amount of keys found, not ones collected.

Reply:
and if we do that, let's move the pattern matching too.

Reply:
sorry, I don't get your concerns, please correct me if i'm wrong. The loop will stop at 'maxiterations' no matter the count of  keys collected, so it will not hang a long time.

Reply:
> and if we do that, let's move the pattern matching too.

Agree and I will try it.

Reply:
the `maxiteration` limit is there in case we're working on an empty dict, one with a lot of empty buckets.
e.g. users sometimes use SCAN+DEL, and then when they re-start scan from 0, all the initial buckets are empty.

but COUNT currently means to stop after that number of keys have been found (not necessarily matched), and if this PR is an efficiency optimization, i don't think we should affect that (i.e. and increase the likelihood we'll hit the `maxiterations` limit.
instead, i think we need to add an `unsigned long sampled` member to `scanData`, increment it inside the scan callback, and then check that the loop instead of checking `listLength(keys)`

Reply:
oh, forgive my stupidity, I finally got what you mean through performance testing...The slowlog shows that when all keys cannot be matched, the scan command of this pr is 5 times slower than the unstable. Thank you very much for your review, which is very very useful!

Reply:
on the ather hand, it scanned ten times the number of keys, while only taking 5 times slower, the performance has been improved by 100% :-)